pilot/pkg/model/sidecar.go

```diff
@@ -766,6 +763,38 @@ func (ilw *IstioEgressListenerWrapper) selectServices(services []*Service, confi
 		}
 	}
 
+	if features.UnifiedSidecarScoping {
+		type namespaceProvider struct {
+			Namespace         string
+			KubernetesService bool
+		}
+		validServices := make(map[host.Name]namespaceProvider, len(importedServices))
+		for _, svc := range importedServices {
+			np := namespaceProvider{Namespace: svc.Attributes.Namespace, KubernetesService: svc.Attributes.ServiceRegistry == provider.Kubernetes}
+			ex, f := validServices[svc.Hostname]
+			// Select a single namespace for a given hostname.
+			// If the same hostname is imported from multiple namespaces, pick the one in the configNamespace
+			// If neither are in configNamespace, an arbitrary one will be chosen, preferring Kubernetes services over ServiceEntry
+			if !f {
+				// First time we saw the hostname, take it
+				validServices[svc.Hostname] = np
+			} else if svc.Attributes.Namespace == configNamespace {
+				// We have seen the hostname, but now its in the configNamespace which has precedence
+				validServices[svc.Hostname] = np
+			} else if !ex.KubernetesService && np.KubernetesService && (np.Namespace == configNamespace || ex.Namespace != configNamespace) {
```

Comment:
I would suggest moving `np.Namespace == configNamespace` handling into the above branch. And maybe `KubernetesService` shoud be a sub branch